File: src/engine.py
"""
engine.py

QuantumOX engine glue: connects Board, Searcher and Options into a single
Engine class suitable for use by main.py (the UTTTI loop).

Responsibilities
- Manage the current Board instance and apply incoming moves
- Expose set_option / new_game / play_move / go functionality
- Convert Searcher results into UTTTI-friendly info lines and bestmove line

Author: Kartik (QuantumOX project)
"""
from __future__ import annotations

from typing import List, Dict, Any, Optional, Tuple

from constants import DEFAULT_GRID
from board import Board
from search import Searcher
from options import set_option, get_option
from utils import (
    format_info_line,
    format_info_string,
    format_bestmove,
)


class QuantumOXEngine:
    def __init__(self) -> None:
        # create board according to current option (if options module not yet configured,
        # fallback to DEFAULT_GRID)
        try:
            grid = get_option("Grid")
        except Exception:
            grid = DEFAULT_GRID
        self.board = Board(grid_spec=grid)
        self.searcher = Searcher()

    # ------------------ options / game lifecycle ---------------------------
    def set_option(self, name: str, value: str) -> Tuple[bool, str]:
        """Set option and perform any side-effects (e.g. changing grid resets board).

        Returns (success, info_message) where info_message is already the text to
        be used in an `info string` UTTTI reply.
        """
        success, msg = set_option(name, value)
        # if Grid changed, recreate board
        if success and name == "Grid":
            # recreate board with new grid spec
            self.board = Board(grid_spec=value)
            # reflect new starting player from options
            try:
                self.board.side_to_move = get_option("FirstPlayer")
            except Exception:
                pass
        return success, msg

    def new_game(self) -> None:
        """Reset board to a fresh game silently (no info string)."""
        self.board = Board(grid_spec=get_option("Grid"))
        try:
            self.board.side_to_move = get_option("FirstPlayer")
        except Exception:
            pass

    def play_moves(self, moves: List[int]) -> str:
        """Apply a sequence of moves to the board (used for `grid ... fill` style commands)."""
        for mv in moves:
            self.board.make_move(mv)
        return format_info_string(f"applied {len(moves)} moves")

    def play_move(self, move: int) -> str:
        """Apply a single move (1-based). Returns an info string confirming the move."""
        self.board.make_move(move)
        return format_info_string(f"played {move}")

    # ------------------ searching / go -----------------------------------
    def go(self, depth: Optional[int] = None, time_ms: Optional[int] = None, nodes: Optional[int] = None) -> Dict[str, Any]:
        """Run search and return a dict with:
            - info_lines: List[str] (UTTTI `info` lines)
            - bestmove_line: str (the final `bestmove ...` line)
            - bestmove: Optional[int]
            - raw: dict (raw search output from Searcher)
        """
        res = self.searcher.search(self.board, max_depth=depth, time_ms=time_ms, nodes_limit=nodes)
        info_lines: List[str] = []

        for d in res.get("infos", []):
            # gather PVs (fall back to empty list if missing)
            pv_main = d.get("pv", [])
            neg_pv = d.get("negamaxpv", [])
            min_pv = d.get("minimaxpv", [])

            # convert to space-separated string of moves
            pv_main_str = " ".join(str(m) for m in pv_main)
            neg_pv_str = " ".join(str(m) for m in neg_pv)
            min_pv_str = " ".join(str(m) for m in min_pv)

            # create a single info line with all three PVs
            line = (
                f"info depth {d.get('depth')} "
                f"seldepth {d.get('seldepth')} "
                f"score cp {d.get('score')} "
                f"nodes {d.get('nodes')} "
                f"minimaxpv {min_pv_str} "
                f"negamaxpv {neg_pv_str} "
                f"pv {pv_main_str}"
            )
            info_lines.append(line)

        bestmove = res.get("bestmove")
        # attempt to set a ponder move to the second PV move if present
        pv = res.get("pv", [])
        ponder = pv[1] if len(pv) > 1 else None
        if bestmove is None:
            bestmove_line = format_bestmove(0, ponder=None)
        else:
            bestmove_line = format_bestmove(bestmove, ponder=ponder)

        return {
            "info_lines": info_lines,
            "bestmove_line": bestmove_line,
            "bestmove": bestmove,
            "raw": res,
        }
    
    def stop(self):
        if hasattr(self, "searcher"):
            self.searcher.abort = True

    # ------------------ utilities ----------------------------------------
    def board_state(self) -> str:
        return str(self.board)

    def legal_moves(self) -> List[int]:
        return self.board.legal_moves()

__all__ = ["QuantumOXEngine"]

File: src/engine.cpp
/* 
 * QuantumOX, a Tic Tac Toe engine supporting UTTTI.
 * Copyright (C) 2025 Kartik Pawar
 *
 * QuantumOX is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * QuantumOX is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. 
 */

#include "engine.h"

#include <string>
#include <vector>
#include <optional>
#include <sstream>
#include <iostream>

namespace QuantumOX {

QuantumOXEngine::QuantumOXEngine() {
    std::string grid;
    try {
        grid = get_option("Grid");
    } catch (...) {
        grid = DEFAULT_GRID;
    }
    board = Board(grid);

    // Attempt to set starting player from options if available
    try {
        std::string fp = get_option("FirstPlayer");
        // board.set_side_to_move(fp); // uncomment if Board exposes setter
    } catch (...) {
        // ignore
    }
}

// ------------------ options / game lifecycle ---------------------------

std::pair<bool,std::string> QuantumOXEngine::set_option(const std::string& name, const std::string& value) {
    auto res = ::QuantumOX::set_option(name, value);
    bool success = res.first;
    std::string msg = res.second;

    if (success && name == "Grid") {
        board = Board(value);
        try {
            std::string fp = get_option("FirstPlayer");
            // board.set_side_to_move(fp);
        } catch(...) {}
    }

    return {success, msg};
}

void QuantumOXEngine::new_game() {
    std::string grid = DEFAULT_GRID;
    try { grid = get_option("Grid"); } catch(...) {}
    board = Board(grid);
    try {
        std::string fp = get_option("FirstPlayer");
        // board.set_side_to_move(fp);
    } catch(...) {}
}

std::string QuantumOXEngine::play_moves(const std::vector<int>& moves) {
    for (int mv : moves) board.make_move(mv);
    return format_info_string("applied " + std::to_string(static_cast<int>(moves.size())) + " moves");
}

std::string QuantumOXEngine::play_move(int move) {
    board.make_move(move);
    return format_info_string("played " + std::to_string(move));
}

// ------------------ search / go ----------------------------------------

GoResult QuantumOXEngine::go(std::optional<int> depth,
                             std::optional<int> time_ms,
                             std::optional<int> nodes) {
    Searcher::SearchResult res = searcher.search(board, depth, time_ms, nodes);

    GoResult out;
    out.raw = res;

    // Build info lines
    // for (const InfoRecord& d : res.infos) {
    //     std::ostringstream oss;
    //     // Build PV strings
    //     std::string pv_main_str, neg_pv_str, min_pv_str;
    //     for (size_t i = 0; i < d.pv.size(); ++i) { if (i) pv_main_str += " "; pv_main_str += std::to_string(d.pv[i]); }
    //     for (size_t i = 0; i < d.negamaxpv.size(); ++i) { if (i) neg_pv_str += " "; neg_pv_str += std::to_string(d.negamaxpv[i]); }
    //     for (size_t i = 0; i < d.minimaxpv.size(); ++i) { if (i) min_pv_str += " "; min_pv_str += std::to_string(d.minimaxpv[i]); }
    // 
    //     oss << "info depth " << d.depth
    //         << " seldepth " << d.seldepth
    //         << " score " << d.score
    //         << " nodes " << d.nodes
    //         << " minimaxpv " << min_pv_str
    //         << " negamaxpv " << neg_pv_str
    //         << " time " << d.time_ms
    //         << " pv " << pv_main_str;
    // 
    //     out.info_lines.push_back(oss.str());
    // }

    // Bestmove
    std::optional<int> bestmove = res.bestmove;
    std::vector<int> pv = res.pv;
    std::optional<int> ponder = std::nullopt;
    if (pv.size() > 1) ponder = pv[1];

    if (!bestmove.has_value()) {
        out.bestmove_line = format_bestmove(0, -1);
        out.bestmove = std::nullopt;
    } else {
        out.bestmove_line = format_bestmove(*bestmove, ponder.has_value() ? *ponder : -1);
        out.bestmove = bestmove;
    }

    return out;
}

// ------------------ stop -----------------------------------------------

void QuantumOXEngine::stop() {
    searcher.request_abort();
}

// ------------------ utilities ------------------------------------------

std::string QuantumOXEngine::board_state() const {
    return board.to_string();
}

std::vector<int> QuantumOXEngine::legal_moves() const {
    return board.legal_moves();
}

} // namespace QuantumOX
